Element application crashes in WebCore::Path::isEmpty()

carlosgcampos opened the pull request:
#### 5ce1ab0db72b993acbfbfc730d3d6f32f6849600
<pre>
Element application crashes in WebCore::Path::isEmpty()
<a href="https://bugs.webkit.org/show_bug.cgi?id=262475">https://bugs.webkit.org/show_bug.cgi?id=262475</a>

Reviewed by Said Abou-Hallawa.

Check hasPath() before calling path().

* Source/WebCore/rendering/svg/RenderSVGPath.cpp:
(WebCore::RenderSVGPath::isRenderingDisabled const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.cpp:
(WebCore::LegacyRenderSVGPath::isRenderingDisabled const):

Canonical link: <a href="https://commits.webkit.org/269223@main">https://commits.webkit.org/269223@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/46764a4dcf235e213967e16458eebd8caf2b86c4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20481 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20907 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21551 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22380 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19112 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20713 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24165 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21082 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20509 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20700 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20566 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17805 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23233 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17734 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18610 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24900 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18811 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18786 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22840 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19381 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16461 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18594 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22934 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2726 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19201 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/rendering/svg/RenderSVGPath.cpp:
```diff
@@ -177,7 +177,7 @@ bool RenderSVGPath::isRenderingDisabled() const
     // For a polygon, polyline or path, rendering is disabled if there is no path data.
     // No path data is possible in the case of a missing or empty 'd' or 'points' attribute.
-    return path().isEmpty();
+    return !hasPath() || path().isEmpty();
```
I saw the assertion in RenderSVGShape::path() in https://bugs.webkit.org/show_bug.cgi?id=149613. And this assertion could have been fixed by the same fix you did. But it turned out to be because of a serious problem that should be fixed somewhere else.

I thought about adding this check but I was afraid this will hide many bugs that will be hard to be discovered. So maybe looking at the test case you are investigating can reveal a serious bug.

carlosgcampos replied:
It's not a simple test unfortunately, it happens to me when loading element app (matrix client). I'll get a backtrace.

nikolaszimmermann replied:
I need a bt ;-)

nikolaszimmermann replied:
I agree with @shallawa we should try hard to understand the exact condition that leads to this.

carlosgcampos replied:
```
#0  0x00007f181edf5b20 in WebCore::Path::isEmpty() const () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#1  0x00007f181f35bbab in WebCore::SVGRenderSupport::computeContainerBoundingBoxes(WebCore::RenderElement const&, WebCore::FloatRect&, bool&, WebCore::FloatRect&, WebCore::FloatRect&) ()
   from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#2  0x00007f181f38197d in WebCore::LegacyRenderSVGRoot::updateCachedBoundaries() () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#3  0x00007f181f381eef in WebCore::LegacyRenderSVGRoot::layout() () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#4  0x00007f181f14d5aa in WebCore::RenderFlexibleBox::layoutAndPlaceChildren(WebCore::LayoutUnit&, WTF::Vector<WebCore::FlexItem, 0ul, WTF::CrashOnOverflow, 16ul, WTF::FastMalloc>&, WebCore::LayoutUnit, bool, WTF::Vector<WebCore::RenderFlexibleBox::LineState, 0ul, WTF::CrashOnOverflow, 16ul, WTF::FastMalloc>&, WebCore::LayoutUnit) ()
   from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#5  0x00007f181f14f134 in WebCore::RenderFlexibleBox::layoutFlexItems(bool) [clone .part.0] () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#6  0x00007f181f14faa9 in WebCore::RenderFlexibleBox::layoutBlock(bool, WebCore::LayoutUnit) () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#7  0x00007f181f0d44b7 in WebCore::RenderBlock::layout() () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#8  0x00007f181f14d5aa in WebCore::RenderFlexibleBox::layoutAndPlaceChildren(WebCore::LayoutUnit&, WTF::Vector<WebCore::FlexItem, 0ul, WTF::CrashOnOverflow, 16ul, WTF::FastMalloc>&, WebCore::LayoutUnit, bool, WTF::Vector<WebCore::RenderFlexibleBox::LineState, 0ul, WTF::CrashOnOverflow, 16ul, WTF::FastMalloc>&, WebCore::LayoutUnit) ()
   from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#9  0x00007f181f14f134 in WebCore::RenderFlexibleBox::layoutFlexItems(bool) [clone .part.0] () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#10 0x00007f181f14faa9 in WebCore::RenderFlexibleBox::layoutBlock(bool, WebCore::LayoutUnit) () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#11 0x00007f181f0d44b7 in WebCore::RenderBlock::layout() () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#12 0x00007f181f10c9c0 in WebCore::RenderBlock::layoutPositionedObject(WebCore::RenderBox&, bool, bool) () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#13 0x00007f181f0c4bbb in WebCore::RenderBlock::layoutPositionedObjects(bool, bool) () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#14 0x00007f181f11ccf7 in WebCore::RenderBlockFlow::layoutBlock(bool, WebCore::LayoutUnit) () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#15 0x00007f181f0d44b7 in WebCore::RenderBlock::layout() () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#16 0x00007f181f10f30a in WebCore::RenderBlockFlow::layoutBlockChild(WebCore::RenderBox&, WebCore::RenderBlockFlow::MarginInfo&, WebCore::LayoutUnit&, WebCore::LayoutUnit&) ()
   from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#17 0x00007f181f11ac2d in WebCore::RenderBlockFlow::layoutBlockChildren(bool, WebCore::LayoutUnit&) () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#18 0x00007f181f11c4de in WebCore::RenderBlockFlow::layoutInFlowChildren(bool, WebCore::LayoutUnit&, WebCore::LayoutUnit&, WebCore::LayoutUnit&) ()
   from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#19 0x00007f181f11ca34 in WebCore::RenderBlockFlow::layoutBlock(bool, WebCore::LayoutUnit) () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#20 0x00007f181f0d44b7 in WebCore::RenderBlock::layout() () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#21 0x00007f181f14eb1e in WebCore::RenderFlexibleBox::alignChildren(WTF::Vector<WebCore::RenderFlexibleBox::LineState, 0ul, WTF::CrashOnOverflow, 16ul, WTF::FastMalloc>&) ()
   from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#22 0x00007f181f14ebd7 in WebCore::RenderFlexibleBox::repositionLogicalHeightDependentFlexItems(WTF::Vector<WebCore::RenderFlexibleBox::LineState, 0ul, WTF::CrashOnOverflow, 16ul, WTF::FastMalloc>&, WebCore::LayoutUnit) () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#23 0x00007f181f14f38d in WebCore::RenderFlexibleBox::layoutFlexItems(bool) [clone .part.0] () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#24 0x00007f181f14faa9 in WebCore::RenderFlexibleBox::layoutBlock(bool, WebCore::LayoutUnit) () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#25 0x00007f181f0d44b7 in WebCore::RenderBlock::layout() () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#26 0x00007f181eb97c5e in WebCore::LocalFrameViewLayoutContext::performLayout() () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#27 0x00007f181eb9815d in WebCore::LocalFrameViewLayoutContext::layout() () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#28 0x00007f181e42146e in WebCore::Document::updateLayout(WTF::OptionSet<WebCore::LayoutOptions>, WebCore::Element const*) () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#29 0x00007f181e43f5a9 in WebCore::Element::offsetTop() () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#30 0x00007f181e455b7e in WebCore::Element::offsetTopForBindings() () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#31 0x00007f181d42a24d in WebCore::jsHTMLElement_offsetTop(JSC::JSGlobalObject*, long, JSC::PropertyName) () from /home/cgarcia/gnome/lib/libwebkitgtk-6.0.so.4
#32 0x00007f181af9579b in JSC::PropertySlot::customGetter(JSC::VM&, JSC::PropertyName) const () from /home/cgarcia/gnome/lib/libjavascriptcoregtk-6.0.so.1
#33 0x00007f181ab05ea3 in JSC::LLInt::performLLIntGetByID(JSC::BytecodeIndex, JSC::CodeBlock*, JSC::JSGlobalObject*, JSC::JSValue, JSC::Identifier const&, JSC::GetByIdModeMetadata&) ()
   from /home/cgarcia/gnome/lib/libjavascriptcoregtk-6.0.so.1
#34 0x00007f181ab06de2 in llint_slow_path_get_by_id () from /home/cgarcia/gnome/lib/libjavascriptcoregtk-6.0.so.1
#35 0x00007f1819d835b9 in llint_op_get_by_id () from /home/cgarcia/gnome/lib/libjavascriptcoregtk-6.0.so.1
#36 0x0000000000000000 in ?? ()
```

Constellation replied:
This `m_path` is nullptr in RenderSVGPath. This means that we have never called `updateShapeFromElement` for this class before. In this condition, using `isRenderingDisabled` is wrong since it is relying on the update from `updateShapeFromElement`. Let's imagine SVG style gets updated (like strokeWidth), and `updateShapeFromElement` needs to be called to reflect this to the path. But if we are not calling this, this `m_path` is pointing at the stale, previous `m_path`, or `nullptr`. And this is the bug here.

When looking at upper container's `LegacyRenderSVGRoot::layout`, it is calling `SVGRenderSupport::layoutChildren`. And we are skipping layout when some conditions are met.
This means that, when computing bounding box, we need to check the exact same condition to skip these non-layout renderers as `isRenderingDisabled`. Current `isRenderingDisabled` is relying on `m_path`, which is only effective after the layout. So, `isRenderingDisabled` implementation looks wrong, this should check both,

1. If layout gets skipped, the same conditions should be used and say `true` from this function.
1. If layout happens, then it should check m_path's emptiness. This is required by SVG's spec IIUC (empty path should not be rendered)

I think checking `hasPath` is not correct here. If we are skipping the layout, then it is checking the stale, previous `m_path` state from the previous layout update.

smfr replied:
We now know that some instances of this crash (e.g. on bing.com) are related to `content-visibility: auto`

carlosgcampos committed:
Element application crashes in WebCore::Path::isEmpty()
https://bugs.webkit.org/show_bug.cgi?id=262475

Reviewed by Said Abou-Hallawa.

Check hasPath() before calling path().

* Source/WebCore/rendering/svg/RenderSVGPath.cpp:
(WebCore::RenderSVGPath::isRenderingDisabled const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.cpp:
(WebCore::LegacyRenderSVGPath::isRenderingDisabled const):

Canonical link: https://commits.webkit.org/269223@main
